src/hnsw/external_index.c

```diff
@@ -547,47 +541,61 @@ void *ldb_wal_index_node_retriever(void *ctxp, int id)
 
     BlockNumber blockno_from_cache = cache_get_item(cache, &id);
     if(blockno_from_cache != InvalidBlockNumber) {
-        blockno = blockno_from_cache;
-    } else {
-        // it is necessary to first check the extra dirtied pages for the blockmap page, in case we are in the
-        // middle of an insert and the insert operation has the block we need under a lock
-        page = extra_dirtied_get(ctx->extra_dirted, blockmapno, NULL);
-        if(page == NULL) {
-            buf = ReadBufferExtended(ctx->index_rel, MAIN_FORKNUM, blockmapno, RBM_NORMAL, NULL);
-            LockBuffer(buf, BUFFER_LOCK_SHARE);
-            page = BufferGetPage(buf);
-        } else {
-            idx_pagemap_prelocked = true;
+        return blockno_from_cache;
+    }
+
+    // it is necessary to first check the extra dirtied pages for the blockmap page, in case we are in the
+    // middle of an insert and the insert operation has the block we need under a lock
+    // if this is th elast page and blocknos is not filled up, only read the part that is filled up
+    // todo:: when we are adding vectors, this does not take into account that the blockmap now has soft-length
+    // of num_vectors + 1
+    Page page = extra_dirtied_get(ctx->extra_dirted, blockmapno, NULL);
+    if(page == NULL) {
+        buf = ReadBufferExtended(ctx->index_rel, MAIN_FORKNUM, blockmapno, RBM_NORMAL, NULL);
+        LockBuffer(buf, BUFFER_LOCK_SHARE);
+        page = BufferGetPage(buf);
+        if(add_to_extra_dirtied) {
+            extra_dirtied_add(ctx->extra_dirted, blockmapno, buf, page);
```

Comment:
Wait, we CANNOT lock the page with `BUFFER_LOCK_SHARE` and end up dirtying it. is something off here?

Reply:
I think you're right. During reads we should acquire share lock, during writes exclusive lock.